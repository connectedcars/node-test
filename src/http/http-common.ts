import http from 'http'

import { Json } from '../common'

export type HttpIncomingMessage = http.IncomingMessage & Required<Pick<http.IncomingMessage, 'method' | 'url'>>

export type HttpRequestListener = (req: HttpIncomingMessage, res: http.ServerResponse) => void

export class HttpServerError extends Error {
  public statusCode = 500
}

export interface HttpRequest {
  id: number
  method: string
  url: string
  headers: http.IncomingHttpHeaders
  body: Buffer
}

export type HttpJsonRequest = Omit<HttpRequest, 'body'> & {
  body: Json
}

export type HttpTextRequest = Omit<HttpRequest, 'body'> & {
  body: string
}

export async function readBody(req: HttpIncomingMessage): Promise<Buffer> {
  return new Promise(resolve => {
    const body: Buffer[] = []
    req.on('data', chunk => {
      body.push(chunk)
    })
    req.on('end', () => {
      resolve(Buffer.concat(body))
    })
  })
}
