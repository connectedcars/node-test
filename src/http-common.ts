import http from 'http'

import { Json } from './common'

export type HttpIncomingMessage = http.IncomingMessage & Required<Pick<http.IncomingMessage, 'method' | 'url'>>

export type HttpRequestListener = (req: HttpIncomingMessage, res: http.ServerResponse) => void

export class HttpServerError extends Error {
  public statusCode = 500
}

export interface HttpRequest {
  method: string
  url: string
  headers: http.IncomingHttpHeaders
  body: Buffer
}

export type HttpJSONRequest = Omit<HttpRequest, 'body'> & {
  body: Json
}

export type HttpStringRequest = Omit<HttpRequest, 'body'> & {
  body: string
}

export async function readBody(req: HttpIncomingMessage): Promise<Buffer> {
  return new Promise(resolve => {
    const body: Buffer[] = []
    req.on('data', chunk => {
      body.push(chunk)
    })
    req.on('end', () => {
      resolve(Buffer.concat(body))
    })
  })
}
