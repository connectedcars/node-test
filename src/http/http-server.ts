import http from 'http'

import { HttpIncomingMessage, HttpRequestListener } from './http-common'
import { HttpServerBase } from './http-server-base'

export type HttpServerOptions = http.ServerOptions & {
  listenPort?: number
  requestIdGenerator?: (req: HttpIncomingMessage) => number
}

export class HttpServer extends HttpServerBase<http.Server> {
  constructor(options: HttpServerOptions, requestListener: HttpRequestListener) {
    super(
      'http://localhost',
      http.createServer(options, (req, res) => {
        const requestId = options.requestIdGenerator
          ? options.requestIdGenerator(req as HttpIncomingMessage)
          : undefined
        this.handleRequest(req as HttpIncomingMessage, res, requestListener, requestId)
      }),
      options.listenPort
    )
  }
}
