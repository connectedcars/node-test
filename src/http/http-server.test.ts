import axios from 'axios'

import { readHttpMessageBody } from './http-common'
import { HttpServer, HttpServerOptions } from './http-server'

export type TestHttpServerOptions = HttpServerOptions

class TestHttpServer extends HttpServer {
  constructor(options: TestHttpServerOptions = {}) {
    super(options, async (req, res) => {
      // Map the responses
      switch (req.url) {
        case '/': {
          return res.end('Hello world')
        }
        case '/json': {
          const body = await readHttpMessageBody(req)
          return res.end(body)
        }
        default: {
          res.statusCode = 404
          return res.end('Not found')
        }
      }
    })
  }
}

describe('HttpServer', () => {
  const httpServer = new TestHttpServer()

  beforeAll(async () => {
    await httpServer.start()
  })

  afterAll(async () => {
    await httpServer.stop()
  })

  afterEach(async () => {
    httpServer.reset()
  })

  it('Simple GET /', async () => {
    const response = await axios.get<string>(`${httpServer.listenUrl}`)
    expect(response.data).toEqual('Hello world')
    expect(httpServer.getTextRequests()).toMatchObject([
      {
        body: '',
        method: 'GET',
        url: '/'
      }
    ])
  })

  it('Simple POST /', async () => {
    const response = await axios.post<string>(`${httpServer.listenUrl}`, 'Hello')
    expect(response.data).toEqual('Hello world')
    expect(httpServer.getTextRequests()).toMatchObject([
      {
        body: 'Hello',
        method: 'POST',
        url: '/'
      }
    ])
    // We have this to test that http requests are stable
    expect(httpServer.getTextRequests()).toMatchSnapshot()
  })

  it('should GET /json and return json requests', async () => {
    const response = await axios.post<string>(`${httpServer.listenUrl}/json`, { test: 'data' })
    expect(response.data).toMatchObject({ test: 'data' })
    expect(httpServer.getJsonRequests()).toMatchObject([
      {
        body: { test: 'data' },
        method: 'POST',
        url: '/json'
      }
    ])
    // We have this to test that http requests are stable
    expect(httpServer.getJsonRequests()).toMatchSnapshot()

    // Validate types
    expect(httpServer.getJsonRequests()[0].body).toMatchObject({ test: 'data' })
    expect(httpServer.getJsonRequests<{ test: string }>()[0].body?.test).toEqual('data')
  })
})
