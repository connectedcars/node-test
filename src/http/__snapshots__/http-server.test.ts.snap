// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpServer Simple POST / 1`] = `
Array [
  Object {
    "body": "Hello",
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "connection": "close",
      "content-length": "5",
      "content-type": "application/x-www-form-urlencoded",
      "host": "localhost",
      "user-agent": "axios/0.19.2",
    },
    "id": 1,
    "method": "POST",
    "url": "/",
  },
]
`;
